(deps): update TypeScript 4.8

As seen in https://github.com/pmndrs/jotai/pull/1369#discussion_r955555431, updating TypeScript 4.8 causes type errors.

- For xstate, we should wait for upstream fix.
- Fix for `jotai/urql` should probably be easy.

#1395 

> * For xstate, we should wait for upstream fix.

Fixed.


> * Fix for `jotai/urql` should probably be easy.



Unfortunately, there are some errors inside `@urql/core`, not in our code.

```
$ tsc --noEmit
node_modules/@urql/core/dist/types/utils/operation.d.ts:2:115 - error TS2344: Type 'Variables' does not satisfy the constraint 'AnyVariables'.
  Type 'Variables' is not assignable to type '{ [prop: string]: any; }'.

2 declare function makeOperation<Data = any, Variables = object>(kind: OperationType, request: GraphQLRequest<Data, Variables>, context: OperationContext): Operation<Data, Variables>;
                                                                                                                    ~~~~~~~~~

  node_modules/@urql/core/dist/types/utils/operation.d.ts:2:44
    2 declare function makeOperation<Data = any, Variables = object>(kind: OperationType, request: GraphQLRequest<Data, Variables>, context: OperationContext): Operation<Data, Variables>;
                                                 ~~~~~~~~~~~~~~~~~~
    This type parameter might need an `extends { [prop: string]: any; }` constraint.
  node_modules/@urql/core/dist/types/utils/operation.d.ts:2:44
    2 declare function makeOperation<Data = any, Variables = object>(kind: OperationType, request: GraphQLRequest<Data, Variables>, context: OperationContext): Operation<Data, Variables>;
                                                 ~~~~~~~~~~~~~~~~~~
    This type parameter might need an `extends AnyVariables` constraint.
```

cc: @kitten

--- a/package.json
+++ b/package.json
@@ -252,10 +252,10 @@
     "ts-expect": "^1.3.0",
     "ts-node": "^10.9.1",
     "tslib": "^2.4.0",
-    "typescript": "4.7.4",
+    "typescript": "^4.8.2",
     "valtio": "^1.6.4",
     "wonka": "^6.0.0",
-    "xstate": "^4.33.4",
+    "xstate": "^4.33.5",
     "zustand": "^4.1.1"
   },
   "resolutions": {

--- a/package.json
+++ b/package.json
@@ -220,7 +220,7 @@
     "@types/react-dom": "^18.0.6",
     "@typescript-eslint/eslint-plugin": "^5.36.0",
     "@typescript-eslint/parser": "^5.36.0",
-    "@urql/core": "^3.0.2",
+    "@urql/core": "^3.0.3",
     "benny": "^3.7.1",
     "concurrently": "^7.3.0",
     "downlevel-dts": "^0.10.1",

--- a/src/urql/atomWithQuery.ts
+++ b/src/urql/atomWithQuery.ts
@@ -17,14 +17,14 @@ type AtomWithQueryAction = {
   opts?: Partial<OperationContext>
 }
 
-type OperationResultWithData<Data, Variables> = OperationResult<
+type OperationResultWithData<
   Data,
-  Variables
-> & {
+  Variables extends AnyVariables
+> = OperationResult<Data, Variables> & {
   data: Data
 }
 
-const isOperationResultWithData = <Data, Variables>(
+const isOperationResultWithData = <Data, Variables extends AnyVariables>(
   result: OperationResult<Data, Variables>
 ): result is OperationResultWithData<Data, Variables> =>
   'data' in result && !result.error

--- a/src/urql/atomWithSubscription.ts
+++ b/src/urql/atomWithSubscription.ts
@@ -11,14 +11,14 @@ import { atom } from 'jotai'
 import type { Atom, Getter } from 'jotai'
 import { clientAtom } from './clientAtom'
 
-type OperationResultWithData<Data, Variables> = OperationResult<
+type OperationResultWithData<
   Data,
-  Variables
-> & {
+  Variables extends AnyVariables
+> = OperationResult<Data, Variables> & {
   data: Data
 }
 
-const isOperationResultWithData = <Data, Variables>(
+const isOperationResultWithData = <Data, Variables extends AnyVariables>(
   result: OperationResult<Data, Variables>
 ): result is OperationResultWithData<Data, Variables> => 'data' in result
 

